File: cpp/api/src/base/TimeCondition.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include "TimeCondition.h"

#include <chrono>
#include <iostream>

namespace cameo {
	
TimeCondition::TimeCondition() :
	m_notified{false} {
}

TimeCondition::~TimeCondition() {
}
	
bool TimeCondition::wait(long timeMs) {

	std::unique_lock<std::mutex> lock {m_conditionMutex};

	if (!m_notified) {
		// Return true if the notification is occurring before the timeout.
		return (m_condition.wait_for(lock, std::chrono::milliseconds{timeMs}) == std::cv_status::no_timeout);
	}
	
	// The notification has occurred.
	return true;
}

void TimeCondition::notify() {

	std::unique_lock<std::mutex> lock {m_conditionMutex};

	m_notified = true;	
	m_condition.notify_one();
}

}

File: java/api/src/main/java/eu/ill/cameo/api/base/TimeCondition.java
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

package eu.ill.cameo.api.base;

/**
 * Class defining a time condition.
 */
public class TimeCondition {

	private boolean notified = false;

	/**
	 * Waits for the time.
	 * @param timeMs Time in milliseconds.
	 * @return True if notification occurred during the wait.
	 */
	public synchronized boolean waitFor(int timeMs) {
		
		try {
			super.wait(timeMs);
		}
		catch (InterruptedException e) {
			return true;
		}
		
		return notified;
	}
	
	/**
	 * Notifies the condition. It unblocks the waitFor() call.
	 */
	public synchronized void notifyCondition() {
		notified = true;
		notify();
	}
}